### ISSUE TITLE
Support relative clangd path

### ISSUE THREAD

**Issue by apockoffork** [clangd/vscode-clangd#22]
My project (which has 100s of easily confused users) has a configured value for `clangd.path`. When I start up vscode now it prompts asking if I want to install clangd. It would be great if the extension could  skip prompting if the path is valid or provide a config option to disable the prompt.

**Comment by hokein** [comment]
just set `"clangd.checkUpdates": false` in the setting.json to disable it.

**Comment by sam-mccall** [comment]
Hmm, no, this sounds more like the autoinstall flow than the auto update flow.

@apockoffork the extension does check if the configured path exists, can you give a little more detail? (Is the configured path the default "clangd" to lookup on PATH, or an absolute path? Does this affect a particular platform? Etc)

**Comment by apockoffork** [comment]
Thanks for the super fast reply. Upon further investigation it seems there is more than one thing probably going on here. 

The `checkUpdates` flag is complaining about the config type
![image](https://user-images.githubusercontent.com/10534342/81582093-9d735d80-937d-11ea-86aa-81925355dd44.png)
That complaint continues on the settings json:
![image](https://user-images.githubusercontent.com/10534342/81582481-1c689600-937e-11ea-80f2-ff5d449408ff.png)

We use a relative path to our clangd executable:
`"clangd.path": "tools/clangd",` 
That wasn't an issue previously but perhaps things have changed and now thats an issue.
The other quirk of our setup is that `tools/clangd` is actually wrapper around the actual clangd executable (it runs like clangd but it isn't technically the clangd binary) so if you're doing some sort of inspection of the file to determine what version of clangd it is or something like that it will probably fail for us.

**Comment by sam-mccall** [comment]
> The checkUpdates flag is complaining about the config type

Oh dear, the default is `"false"` instead of `false`, which means... true :-(
I've pushed 0.1.3 to fix this. Thanks!

> We use a relative path to our clangd executable

Ah, and this is relative to `$PATH`, rather than the working directory?
I'm not surprised this doesn't work after the patch (but a little surprised that it worked before). 
I don't think this is standard, e.g. `bash` doesn't seem to accept this:

```
sammccall@pope:~$ bin/xx/foo.sh
foosh
sammccall@pope:~$ PATH=$PWD/bin/xx
sammccall@pope:~$ foo.sh
foosh
sammccall@pope:~$ PATH=$PWD/bin
sammccall@pope:~$ xx/foo.sh
bash: xx/foo.sh: No such file or directory
```

Is it possible to use either an absolute path, or a file directly on `$PATH`?

> if you're doing some sort of inspection of the file to determine what version of clangd it is or something like that it will probably fail for us

We do three things with the binary, as of 0.1.2:
 - we check whether it exists (relative to `$PATH`, with optional `$PATHEXT`, using the npm package [which](https://www.npmjs.com/package/which)). If not, we prompt to install. I think this is the check that's failing.
 - If updates are enabled, we run it with `--version` to see if it should be upgraded. If we can't parse the result then the answer is no, it's not upgradable. So this should be OK. (If you want to disable autoupdates, you can either force the setting `clangd.checkUpdates` to false, or make sure the output doesn't start with `clangd version X.Y.Z`. e.g. our internal distribution is `clangd version google-trunk`.
 - We run it as the actual language server. (This should be unchanged from previous versions).

So I think a wrapper per se shouldn't be a problem.

**Comment by apockoffork** [comment]
Unfortunately we can't use an absolute path for a global config though we can direct our users to set one. I tested it and verified that does solve the issue. 

We have a "fully" hermetic bazel build system implemented so the clang tool chain is all built and available at a path relative to the repo root which is going to vary user to user. For the majority of our extensions we try to make use of the `${workspaceRoot}` config variable but support for that seems to be extension dependent. For example our python interpreter path is:
`"python.pythonPath": "${workspaceRoot}/tools/python3_ide",`.

**Comment by sam-mccall** [comment]
> For the majority of our extensions we try to make use of the ${workspaceRoot} config variable but support for that seems to be extension dependent

It's probably best for us to support that, I'll put together a patch ASAP (though it's evening here and might need to get the kids to bed first!).

> Also wanted to take a moment just to thank all the awesome work you guys have been doing on clangd

That's very gracious of you, especially after we just broke everything for you. I'm glad things are working well :-)

**Comment by jvilk-stripe** [comment]
Thanks for the quick fix! :) I was having the same issue with our [vscode-clangd configuration for Sorbet](https://github.com/sorbet/sorbet/blob/master/.vscode/settings.json#L11); we use bazel to pull in clangd, which adds it to a subdirectory of the workspace root.

I look forward to the next version with the fix!

**Comment by sam-mccall** [comment]
I think this is fixed in 0.1.4:
 - we support substitution of `${workspaceRoot}/foo/bar`, which I would recommend using. `${env:XYZ}` and `${config:foo.bar}` should also work, most other substitutions do not.
 - `tools/clangd` does *approximately* what it did before: it will be resolved relative to the workspace root if there is one. This legacy behavior is only triggered if there's a slash in the path.

Please let us us know if you still have problems.

**Comment by apockoffork** [comment]
Thanks for the quick fix. All better now!
